pkg/block/fetcher.go

```diff
@@ -785,25 +787,69 @@ func (f *IgnoreDeletionMarkFilter) DeletionMarkBlocks() map[ulid.ULID]*metadata.
 func (f *IgnoreDeletionMarkFilter) Filter(ctx context.Context, metas map[ulid.ULID]*metadata.Meta, synced *extprom.TxGaugeVec) error {
 	f.deletionMarkMap = make(map[ulid.ULID]*metadata.DeletionMark)
 
+	// Make a copy of block IDs to check, in order to avoid concurrency issues
+	// between the scheduler and workers.
+	blockIDs := make([]ulid.ULID, 0, len(metas))
 	for id := range metas {
-		m := &metadata.DeletionMark{}
-		if err := metadata.ReadMarker(ctx, f.logger, f.bkt, id.String(), m); err != nil {
-			if errors.Cause(err) == metadata.ErrorMarkerNotFound {
-				continue
+		blockIDs = append(blockIDs, id)
+	}
+
+	var (
+		eg  errgroup.Group
+		ch  = make(chan ulid.ULID, f.concurrency)
+		mtx sync.Mutex
+	)
+
+	for i := 0; i < f.concurrency; i++ {
+		eg.Go(func() error {
+			for id := range ch {
+				m := &metadata.DeletionMark{}
+				if err := metadata.ReadMarker(ctx, f.logger, f.bkt, id.String(), m); err != nil {
+					if errors.Cause(err) == metadata.ErrorMarkerNotFound {
+						continue
+					}
+					if errors.Cause(err) == metadata.ErrorUnmarshalMarker {
+						level.Warn(f.logger).Log("msg", "found partial deletion-mark.json; if we will see it happening often for the same block, consider manually deleting deletion-mark.json from the object storage", "block", id, "err", err)
+						continue
+					}
+					return err
+				}
+
+				// Keep track of the blocks marked for deletion and filter them out if their
+				// deletion time is greater than the configured delay.
+				mtx.Lock()
+				f.deletionMarkMap[id] = m
+				if time.Since(time.Unix(m.DeletionTime, 0)).Seconds() > f.delay.Seconds() {
+					synced.WithLabelValues(markedForDeletionMeta).Inc()
+					delete(metas, id)
+				}
+				mtx.Unlock()
 			}
-			if errors.Cause(err) == metadata.ErrorUnmarshalMarker {
-				level.Warn(f.logger).Log("msg", "found partial deletion-mark.json; if we will see it happening often for the same block, consider manually deleting deletion-mark.json from the object storage", "block", id, "err", err)
-				continue
+
+			return nil
+		})
+	}
+
+	// Workers scheduled, distribute blocks.
+	eg.Go(func() error {
+		defer close(ch)
+
+		for _, id := range blockIDs {
+			select {
+			case ch <- id:
+				// Nothing to do.
+			case <-ctx.Done():
+				return ctx.Err()
 			}
-			return err
 		}
 
-		f.deletionMarkMap[id] = m
-		if time.Since(time.Unix(m.DeletionTime, 0)).Seconds() > f.delay.Seconds() {
-			synced.WithLabelValues(markedForDeletionMeta).Inc()
-			delete(metas, id)
-		}
+		return nil
+	})
+
+	if err := eg.Wait(); err != nil {
```

Comment:
And if writing to the channel blocked then ```eg.Wait()``` would not return as it waits for all go routines to finish?

The simplest mitigation would be to change capacity of the channel from f.Concurrency to len(blockIDs)?